### PR TITLE
proj search path to manually add transformation grid files

### PR DESCRIPTION
**Pull request by nirvn**
This PR does quite a few things:
- it adds to proj a new searchable directory, <external_storage>/QField/proj , where users can drop in grid files
- it adds a UI handler to QField to provide better feedback to users when grids are missing. Hyperlink downloads to known grids is provided in the log popup when available.
- it also adds a new settings panels, GRIDS, where a user can select a grid to be used to adjust vertical (i.e. Z) value of the GPS location

Screenshot of new settings UI:
image

@m-kuhn , I think this is ready for round one of review.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-09df69215a82c5cc9bdab2237dfda75024d2a800-arm64_v8a.apk)

Other architectures:

  - [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-09df69215a82c5cc9bdab2237dfda75024d2a800-armv7.apk)
  - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-09df69215a82c5cc9bdab2237dfda75024d2a800-x86_64.apk)
  - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-09df69215a82c5cc9bdab2237dfda75024d2a800-x86.apk)

**Comment by qfield-fairy** [comment]
Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-aef51b5baa4ca41862250646f843664337d11db9-arm64_v8a.apk)

Other architectures:

  - [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-aef51b5baa4ca41862250646f843664337d11db9-armv7.apk)
  - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-aef51b5baa4ca41862250646f843664337d11db9-x86_64.apk)
  - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-aef51b5baa4ca41862250646f843664337d11db9-x86.apk)

**Comment by qfield-fairy** [comment]
Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-dda7f7d94d9e1a5fd56a7df2f8bb59cca88841e0-arm64_v8a.apk)

Other architectures:

  - [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-dda7f7d94d9e1a5fd56a7df2f8bb59cca88841e0-armv7.apk)
  - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-dda7f7d94d9e1a5fd56a7df2f8bb59cca88841e0-x86_64.apk)
  - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-dda7f7d94d9e1a5fd56a7df2f8bb59cca88841e0-x86.apk)

File: src/core/appcoordinateoperationhandlers.cpp [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    if ( !grid.packageName.isEmpty() )
+    {
+      downloadMessage = tr( "This grid is part of the “<i>%1</i>” package, available for download from <a href=\"%2\">%2</a>." ).arg( grid.packageName, grid.url );
+    }
+    else
+    {
+      downloadMessage = tr( "This grid is available for download from <a href=\"%1\">%1</a>." ).arg( grid.url );
+    }
+  }
+  QgsMessageLog::logMessage( QStringLiteral( "%1\n%2"  ).arg( shortMessage, downloadMessage ), QStringLiteral( "projection" ) );
```
Comment (m-kuhn) [inline src/core/appcoordinateoperationhandlers.cpp:97]:
```suggestion
  QgsMessageLog::logMessage( QStringLiteral( "%1\n%2"  ).arg( shortMessage, downloadMessage ), QStringLiteral( tr( "projection" ) ) );
```

?

Comment (m-kuhn) [reply src/core/appcoordinateoperationhandlers.cpp:97]: Also, should we add a link to the documentation that explains how to install these files into the correct location?

**Comment by qfield-fairy** [comment]
Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-85f7039f65b84a947703b852d01aac2092caa55b-arm64_v8a.apk)

Other architectures:

  - [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-85f7039f65b84a947703b852d01aac2092caa55b-armv7.apk)
  - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-85f7039f65b84a947703b852d01aac2092caa55b-x86_64.apk)
  - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-85f7039f65b84a947703b852d01aac2092caa55b-x86.apk)

**Comment by qfield-fairy** [comment]
Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-88e6b29140d689d556cbf0a8a2488167446a362c-arm64_v8a.apk)

Other architectures:

  - [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-88e6b29140d689d556cbf0a8a2488167446a362c-armv7.apk)
  - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-88e6b29140d689d556cbf0a8a2488167446a362c-x86_64.apk)
  - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-88e6b29140d689d556cbf0a8a2488167446a362c-x86.apk)

File: src/core/platformutilities.cpp [hunk 2]
```diff
@@ -51,11 +52,24 @@ QString PlatformUtilities::qgsProject() const
...
 }
 
+QStringList PlatformUtilities::availableGrids() const
+{
+  if ( !qfieldDataDir().isEmpty() )
+  {
+    QDir gridsDir( qfieldDataDir() + "proj/" );
+    if ( gridsDir.exists() )
+    {
+      return gridsDir.entryList( QStringList() << QStringLiteral( "*.tif" ) << QStringLiteral( "*.gtx" ) << QStringLiteral( "*.gsb" ) );
```
Comment (m-kuhn) [inline src/core/platformutilities.cpp:67]:
```suggestion
      return gridsDir.entryList( QStringList() << QStringLiteral( "*.tif" ) << QStringLiteral( "*.gtx" ) << QStringLiteral( "*.gsb" ) << QStringLiteral( "*.byn" ) );
```

**Comment by qfield-fairy** [comment]
Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-6102416a56afdcfad58f9632bb1281c51f55ee75-arm64_v8a.apk)

Other architectures:

  - [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-6102416a56afdcfad58f9632bb1281c51f55ee75-armv7.apk)
  - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-6102416a56afdcfad58f9632bb1281c51f55ee75-x86_64.apk)
  - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-6102416a56afdcfad58f9632bb1281c51f55ee75-x86.apk)

**Comment by qfield-fairy** [comment]
Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-3827a9541da64fd231f1b0fdf76c2ed240ddfc1c-arm64_v8a.apk)

Other architectures:

  - [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-3827a9541da64fd231f1b0fdf76c2ed240ddfc1c-armv7.apk)
  - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-3827a9541da64fd231f1b0fdf76c2ed240ddfc1c-x86_64.apk)
  - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-3827a9541da64fd231f1b0fdf76c2ed240ddfc1c-x86.apk)

**Comment by qfield-fairy** [comment]
Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-e1d7be3f591a14a2c5a5f404b2e84df0deb104e8-arm64_v8a.apk)

Other architectures:

  - [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-e1d7be3f591a14a2c5a5f404b2e84df0deb104e8-armv7.apk)
  - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-e1d7be3f591a14a2c5a5f404b2e84df0deb104e8-x86_64.apk)
  - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-e1d7be3f591a14a2c5a5f404b2e84df0deb104e8-x86.apk)

File: src/core/qgsquick/qgsquickcoordinatetransformer.cpp [hunk 3]
```diff
@@ -181,14 +181,14 @@ void QgsQuickCoordinateTransformer::updatePosition()
...
-    std::vector< double > yVector = { mSourcePosition.y() };
-    std::vector< double > zVector = { !std::isnan( mSourcePosition.z() ) ? mSourcePosition.z() : 0 };
+    std::vector< double > xVector = { 7.9995 };//mSourcePosition.x() };
+    std::vector< double > yVector = { 46.7949 };//mSourcePosition.y() };
+    std::vector< double > zVector = { 10 };//!std::isnan( mSourcePosition.z() ) ? mSourcePosition.z() : 0 };
     double zDummy = 0.0; // we don't want to manipulate the elevation data yet, use a dummy z value to transform coordinates first
     mCoordinateVerticalGridTransform.transformInPlace( xVector[0], yVector[0], zDummy );
 
     PJ *P = proj_create( PJ_DEFAULT_CTX, QStringLiteral( "+proj=pipeline +step +proj=unitconvert +xy_in=deg +xy_out=rad +step +proj=vgridshift +grids=%1 +step +proj=unitconvert +xy_in=rad +xy_out=deg" ).arg( verticalGridName ).toUtf8().constData() );
-    proj_trans_generic( P, PJ_FWD,
```
Comment (m-kuhn) [inline src/core/qgsquick/qgsquickcoordinatetransformer.cpp]: My test file here required `PJ_FWD` (i.e. works smoothly before this commit), what was the reason for this change?

**Comment by qfield-fairy** [comment]
Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-0c71ceff6214af42c03606f63883ac2fd3e1ee17-arm64_v8a.apk)

Other architectures:

  - [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-0c71ceff6214af42c03606f63883ac2fd3e1ee17-armv7.apk)
  - [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-0c71ceff6214af42c03606f63883ac2fd3e1ee17-x86_64.apk)
  - [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-0c71ceff6214af42c03606f63883ac2fd3e1ee17-x86.apk)

File: src/core/qgsquick/qgsquickcoordinatetransformer.cpp (hunk 3 above)
Comment (nirvn) [reply src/core/qgsquick/qgsquickcoordinatetransformer.cpp]: Clarified and adapted.

File: src/core/qgsquick/qgsquickcoordinatetransformer.cpp [hunk 4]
```diff
@@ -14,12 +14,29 @@
  ***************************************************************************/
 
 #include "qgsquickcoordinatetransformer.h"
+#include "platformutilities.h"
+#if defined(Q_OS_ANDROID)
+#include "androidplatformutilities.h"
+#endif
+
+#include <QDebug>
```
Comment (m-kuhn) [inline src/core/qgsquick/qgsquickcoordinatetransformer.cpp:22]:
```suggestion
```

File: src/core/qgsquick/qgsquickcoordinatetransformer.cpp [hunk 5]
```diff
@@ -127,6 +144,62 @@ void QgsQuickCoordinateTransformer::updatePosition()
   if ( mSkipAltitudeTransformation )
     z = mSourcePosition.z();
 
+  QSettings settings;
+  const QString verticalGridName = settings.value( QStringLiteral( "verticalGrid" ), QString() ).toString();
+  if ( !verticalGridName.isEmpty () )
+  {
+    if ( mVerticalGridName != verticalGridName )
+    {
+#if defined(Q_OS_ANDROID)
```
Comment (m-kuhn) [inline src/core/qgsquick/qgsquickcoordinatetransformer.cpp:153]: Hmmm... that needs a bit of reworking.

Currently we have platformUtils defined in qgismobileapp.cpp. This needs to be changed.

```

#if defined (Q_OS_ANDROID)
Q_GLOBAL_STATIC(AndroidPlatformUtilities, sPlatformUtils)
#else ...
...
#endif

PlatformUtilities *PlatformUtilities::instance()
{
  return sPlatformUtils;
}
```

And then always use `PlatformUtilities::instance()` everywhere. This way we can register new platforms at a single point in the source code.

File: src/qml/QFieldSettings.qml [hunk 6]
```diff
@@ -166,8 +174,49 @@ Page {
...
+                  text: qsTr( "Select a vertical grid shift in the combo box below to increase vertical location accuracy. Leave blank to disable this feature." )
+                  font: Theme.defaultFont
+
+                  wrapMode: Text.WordWrap
+                  Layout.fillWidth: true
+              }
+
+              ComboBox {
+                  Layout.fillWidth: true
+                  model: [ "" ].concat( platformUtilities.availableGrids() );
```
Comment (m-kuhn) [inline src/qml/QFieldSettings.qml:195]: Can we add a text here like "<i>Use raw altitude from positioning device</i>" or similar?

Comment (nirvn) [reply src/qml/QFieldSettings.qml:195]: @m-kuhn , what about 'No grid (use raw altitude from device)'?

**Comment by nirvn** [comment]
@m-kuhn , review addressed.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-70d79ae5b1149858a508e7e49b14dcd60f654a82-arm64_v8a.apk)

Other architectures: [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-70d79ae5b1149858a508e7e49b14dcd60f654a82-armv7.apk), [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-70d79ae5b1149858a508e7e49b14dcd60f654a82-x86_64.apk), [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-70d79ae5b1149858a508e7e49b14dcd60f654a82-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-c554f164b0a1b533dcab3827861662b4b020eb67-arm64_v8a.apk)

Other architectures: [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-c554f164b0a1b533dcab3827861662b4b020eb67-armv7.apk), [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-c554f164b0a1b533dcab3827861662b4b020eb67-x86_64.apk), [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-c554f164b0a1b533dcab3827861662b4b020eb67-x86.apk)

File: src/core/qgismobileapp.cpp [hunk 7]
```diff
@@ -198,6 +236,8 @@ QgisMobileapp::QgisMobileapp( QgsApplication *app, QObject *parent )
   mOfflineEditing = new QgsOfflineEditing();
 
   mSettings.setValue( "/Map/searchRadiusMM", 5 );
+
+  ( void )new AppMissingGridHandler( this );
```
Comment (m-kuhn) [inline src/core/qgismobileapp.cpp:240]: Can we store this pointer somewhere (e.g. in a member variable)?
Sooner or later we'll want to react to missing grids, probably by connecting to signals from this handler.

**Review by m-kuhn — COMMENTED** [review]
Thanks for those adjustments !

**Review by m-kuhn — APPROVED** [review]

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-8fff36e326d5779ddda8ebf4440da1da4e34be25-arm64_v8a.apk)

Other architectures: [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-8fff36e326d5779ddda8ebf4440da1da4e34be25-armv7.apk), [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-8fff36e326d5779ddda8ebf4440da1da4e34be25-x86_64.apk), [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1385-8fff36e326d5779ddda8ebf4440da1da4e34be25-x86.apk)

File: src/core/qgismobileapp.cpp (hunk 7 above)
Comment (nirvn) [reply src/core/qgismobileapp.cpp]: Done.

File: CMakeLists.txt [hunk 8]
```diff
@@ -25,6 +25,8 @@ CMAKE_POLICY(SET CMP0071 NEW)
 
 ENABLE_TESTING()
 FIND_PACKAGE(QGIS)
+FIND_PACKAGE(Proj)
```
Comment (3nids) [inline CMakeLists.txt:28]: shouldn't this be REQUIRED?

Comment (nirvn) [reply CMakeLists.txt:28]: Yes it should.

**Pull request merged**